rag/llm/key_utils.py

```
@@ -31,4 +34,72 @@ def _normalize_replicate_key(key):
+        return key
+    if isinstance(key, str):
+        try:
+            payload = json.loads(key)
+        except (json.JSONDecodeError, TypeError):
+            logging.warning(
+                'OCR provider key is not valid JSON; expected a JSON object (e.g. nested {"api_key": {...}} from the UI, or flat {"PROVIDER_*": "..."} auto-provisioned from env vars). See conf/llm_factories.json for supported OCR factories.',
+            )
+            raise ModelException(
+                'OCR provider key must be a JSON object. Example: {"api_key": {"mineru_apiserver": "..."}} or {"MINERU_API_KEY": "..."}. See conf/llm_factories.json for supported OCR factories.',
+                retryable=False,
+            )
```

Ruff (0.16.0)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)